File: py_drill/Problems/TR02_SameTree.py
# REF : https://leetcode.com/problems/same-tree/
# NOTES :
#       Base condition - If both p, q are null then they are same -> True
#       While traversing any of the node is None then return False
#       Check if the node value is not same then return False
#       call recursively to left right for both

# T: O(n)
# S: O(n) stack space
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSameTree(self, p, q) -> bool:
        if not p and not q:
            return True

        if not p or not q:
            return False

        if p.val != q.val:
            return False

        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

        # Also same logic

        # if p and q and p.val == q.val:
        #     return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        # return False



def main():
    pass

if __name__ == "__main__":
    main()

File: cpp_drill/Problems/TR02_SameTree.cpp
#include <iostream>
#include <string>
using namespace std;

// REF : https://leetcode.com/problems/same-tree/
// NOTES :
//     Base condition - If both p, q are null then they are same -> True
//     While traversing any of the node is None then return False
//     Check if the node value is not same then return False
//     call recursively to left right for both
// T: O(n)
// S: O(n) stack space

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == nullptr && q == nullptr)
            return true;
        
        if (p == nullptr || q == nullptr)
            return false;

        return p->val == q->val &&
                isSameTree(p->left, q->left) &&
                isSameTree(p->right, q->right);
    }
};

int main() {
    return 0;
}